src/zeemandata.cc

```diff
@@ -147,6 +157,49 @@ Zeeman::Model Zeeman::GetAdvancedModel(const QuantumIdentifier& qid) noexcept {
         return Model({gu, gl});
       }
     }
+  } else if (qid.SpeciesName() == "CO") {
+    if (qid.Isotopologue() == SpeciesTag("CO-26").Isotopologue()) {
+      Numeric gperp = -0.2689 / Constant::mass_ratio_electrons_per_proton;  // Flygare and Benson 1971
+      
+      return Model({gperp, gperp});
+    }
+  } else if (qid.SpeciesName() == "OCS") {
+    if (qid.Isotopologue() == SpeciesTag("OCS-622").Isotopologue()) {
+      Numeric gperp = -.02889 / Constant::mass_ratio_electrons_per_proton;  // Flygare and Benson 1971
+      Numeric gpara = 0 / Constant::mass_ratio_electrons_per_proton;  // Flygare and Benson 1971
+      
+      auto JU = qid.UpperQuantumNumber(QuantumNumberType::J);
+      auto KU = qid.UpperQuantumNumber(QuantumNumberType::Ka);
+      auto JL = qid.LowerQuantumNumber(QuantumNumberType::J);
+      auto KL = qid.LowerQuantumNumber(QuantumNumberType::Ka);
+      
+      return Model({closed_shell_trilinear(KU, JU, gperp, gpara),
+                    closed_shell_trilinear(KL, JL, gperp, gpara)});
+    } else if (qid.Isotopologue() == SpeciesTag("OCS-624").Isotopologue()) {
+      Numeric gperp = -.0285 / Constant::mass_ratio_electrons_per_proton;  // Flygare and Benson 1971
+      Numeric gpara = -.061 / Constant::mass_ratio_electrons_per_proton;  // Flygare and Benson 1971
+      
+      auto JU = qid.UpperQuantumNumber(QuantumNumberType::J);
+      auto KU = qid.UpperQuantumNumber(QuantumNumberType::Ka);
+      auto JL = qid.LowerQuantumNumber(QuantumNumberType::J);
+      auto KL = qid.LowerQuantumNumber(QuantumNumberType::Ka);
+      
+      return Model({closed_shell_trilinear(KU, JU, gperp, gpara),
+                    closed_shell_trilinear(KL, JL, gperp, gpara)});
```

Comment:
Don't see any way to handle that more cleanly for all species in general due to the very different parameters needed for each species. But we can try to reduce the redundant code per isotopologue. Something like this maybe:

```cpp
 } else if (qid.SpeciesName() == "OCS") {
    Numeric gperp, gpara;

    if (qid.Isotopologue() == SpeciesTag("OCS-622").Isotopologue()) {
      gperp = -.02889 / Constant::mass_ratio_electrons_per_proton;  // Flygare and Benson 1971
      gpara = 0 / Constant::mass_ratio_electrons_per_proton;  // Flygare and Benson 1971
    } else if (qid.Isotopologue() == SpeciesTag("OCS-624").Isotopologue()) {
      gperp = -.0285 / Constant::mass_ratio_electrons_per_proton;  // Flygare and Benson 1971
      gpara = -.061 / Constant::mass_ratio_electrons_per_proton;  // Flygare and Benson 1971
    } else {
      return Model();
    }
    const auto JU = qid.UpperQuantumNumber(QuantumNumberType::J);
    const auto KU = qid.UpperQuantumNumber(QuantumNumberType::Ka);
    const auto JL = qid.LowerQuantumNumber(QuantumNumberType::J);
    const auto KL = qid.LowerQuantumNumber(QuantumNumberType::Ka);

    return Model({closed_shell_trilinear(KU, JU, gperp, gpara),
                  closed_shell_trilinear(KL, JL, gperp, gpara)});
```

Reply:
I like the clearer split; that there is either a zero-return or a good return in the end.  Your example saves some lines but I find it more cumbersome to deal with since it would have to happen anyways per species.

That said, I don't like that the only way I can figure out how to get to the 12C-16O isotopologue, for example, is by the use of SpeciesTag("CO-26").  There must be some constexpr-able way of having the Index of the isotopologue automatically found out so that an overarching switch-statement for all isotopologue can be used.  Especially as I am almost certain that "CO-26" is the same isotopologue in ARTS as "OCS-622".  We would be better off separating the real isotopologue and the full models with a redesign of species data...

Reply:
I'm fine with leaving it as it is. With a redesign of species_data we should hold off until more of the data types actually support constexpr. Since there is no constexpr support for enums, strings or vectors yet, it feels like quite some compromises in the implementation would be needed to get it to work right now. And while it would be convenient to be able to use a switch statement instead of if-elseif..., the gain would be rather marginally to justify a complete redesign.

Should I go ahead with merging this PR?

Reply:
Right, I agree we can wait until there are more constexpr things.  It's just that I am working with quite a few things that require species-specific data that is not in species_data.  It would be nice to have a cleaner way to define species_data so that these species data can be modified and added.  For instance, the numbers in for O2 Zeeman effect are what is needed to compute the energy levels for line mixing.  These are therefore defined twice, once per functionality.  Since I know the numbers are good for the Zeeman stuff, they should be the same for line mixing stuff.  Yet because the line mixing stuff is already so complicated and there is still a bit to do to make it work (and I have little time for it), I have had to redefine things to make the functions readable.

Anyways, it is OK to go ahead and merge this code.  I will see if I can update other things later.